File: src/ros2_bridge/mappers/control_msgs/joint_jog.rs
//! Typed mapper for `control_msgs/msg/JointJog`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn joint_jog_to_bus(msg: ros_env::control_msgs::msg::JointJog) -> crate::control_msgs::msg::v1::JointJog {
    crate::control_msgs::msg::v1::JointJog {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        joint_names: crate::ros2_bridge::mappers::convert::string_seq(msg.joint_names),
        displacements: crate::ros2_bridge::mappers::convert::f64_seq(msg.displacements),
        velocities: crate::ros2_bridge::mappers::convert::f64_seq(msg.velocities),
        duration: msg.duration,
    }
}

pub(crate) fn joint_jog_to_ros(bus: crate::control_msgs::msg::v1::JointJog) -> ros_env::control_msgs::msg::JointJog {
    ros_env::control_msgs::msg::JointJog {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        joint_names: crate::ros2_bridge::mappers::convert::ros_string_seq(bus.joint_names),
        displacements: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.displacements),
        velocities: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.velocities),
        duration: bus.duration,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct ControlMsgsJointJogMapper;

impl TypedTopicMapper for ControlMsgsJointJogMapper {
    type Ros = ros_env::control_msgs::msg::JointJog;
    type Bus = crate::control_msgs::msg::v1::JointJog;

    fn type_name(&self) -> &'static str {
        "control_msgs/msg/JointJog"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(joint_jog_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(joint_jog_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/control_msgs/joint_jog.py
"""Generated mapper for `control_msgs/msg/JointJog`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.control_msgs.msg.v1 import JointJog as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def joint_jog_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.joint_names.extend([str(x) for x in msg.joint_names])
    bus.displacements.extend(list(msg.displacements))
    bus.velocities.extend(list(msg.velocities))
    bus.duration = msg.duration
    return bus


def joint_jog_to_ros(bus):
    from control_msgs.msg import JointJog as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.joint_names = [str(x) for x in bus.joint_names]
    out.displacements = list(bus.displacements)
    out.velocities = list(bus.velocities)
    out.duration = bus.duration
    return out


class ControlMsgsJointJogMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from control_msgs.msg import JointJog as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return joint_jog_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return joint_jog_to_ros(bus)
